File: 1_two_sum/two_sum.go
package main

import (
	"fmt"
)

func twoSum(nums []int, target int) []int {
	difMap := make(map[int]int)

	for ind, val := range nums {
		difMap[val] = ind
	}

	result := []int{0, 1}
	for ind, val := range nums {
		lookup := target - val

		pos, ok := difMap[lookup]

		if ok {

			if pos == ind {
				continue
			} else {
				result = []int{ind, pos}
				break
			}
		}
	}

	return result
}

func main() {
	fmt.Println("Potato")

	// Input: nums = [2,7,11,15], target = 9
	// Output: [0,1]
	nums := []int{2, 7, 11, 15}
	fmt.Println(twoSum(nums, 9))

	// Input: nums = [3,2,4], target = 6
	// Output: [1,2]
	nums = []int{3, 2, 4}
	fmt.Println(twoSum(nums, 6))

	// Input: nums = [3,3], target = 6
	// Output: [0,1]
	nums = []int{3, 3}
	fmt.Println(twoSum(nums, 6))

	// Input: nums = [3,2,3], target = 6
	// Output: [0,1]
	nums = []int{3, 2, 3}
	fmt.Println(twoSum(nums, 6))

}

File: 1_two_sum/two_sum.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hashMap = {}

        for key, val in enumerate(nums):
            hashMap[val] = key
        
        for key2, val in enumerate(nums):
            if target-val in hashMap.keys() and key2 != hashMap[target-val]:
                return [key2, hashMap[target-val]]

        

if __name__ == "__main__":
    sol = Solution()
    nums = [2,7,11,15]
    target = 9
    print( sol.twoSum(nums, target))

    nums = [3,2,4]
    target = 6
    print( sol.twoSum(nums, target))

    nums = [3,3]
    target = 6
    print( sol.twoSum(nums, target))
